### PR TITLE
Better integration with camel

### LINKED ISSUE #121: Better Integration with camel
**As** a user of the extension ` io.quarkiverse:quarkus-artemis-jms` who also uses camel

**when** I define `ConnectionFactory`s via a configuration file (`application.properties`/`application.yml`)

**then** I want that that those `ConnectionFactory`s are automatically registered in the camel context, so that I can use them via their name, e.g.:
```
from(jms("queue:in").connectionFactory("<default>"))
    .to(jms("queue:out").connectionFactory("named"));
```

---

**As** a user of the extension ` io.quarkiverse:quarkus-artemis-jms` who also uses camel

**when** I define `ConnectionFactory`s as beans in my application that have an [`@io.smallrye.common.annotation.Identifier("...")`](https://github.com/smallrye/smallrye-common/blob/main/annotation/src/main/java/io/smallrye/common/annotation/Identifier.java)

**then** I want that that those `ConnectionFactory`s are automatically registered in the camel context, so that I can use them via their name, e.g.:
```
from(jms("queue:in").connectionFactory("named"))
    .to(jms("queue:out").connectionFactory("externally-defined"));
```

---

**As** a user of the extension ` io.quarkiverse:quarkus-artemis-jms` who also uses camel

**when** I use only one `ConnectionFactory` (either through configuration or externally defined)

**then** I want that this connection factory is registered automatically, and I do not have to set it explicitly, e.g.:
```
from(jms("queue:in"))
    .to(jms("queue:out"));
```

### PR DESCRIPTION
**Pull request by turing85**
Resolves #121

### CODE REVIEW

**Review by zhfeng — CHANGES_REQUESTED** [review]
I'd like to hold this PR and run the tests with `camel-quarkus` to see if they are working. If there is any failure, we can try to figure it out in `camel-quarkus` at first.

**Comment by turing85** [comment]
I have validated this build locally against `quarkus-pooled-jms`, and everything works as expected.

**Comment by zhfeng** [comment]
@turing85  That's great and is it possible to add these tests on https://github.com/apache/camel-quarkus/tree/main/integration-tests/jms-artemis-client ?

**Comment by turing85** [comment]
Looking at the tests you linked (I actually used them as template for my tests), isn't this case covered by the [`with-external`-tests](https://github.com/quarkiverse/quarkus-artemis/tree/feature/121-better-integration-with-camel/integration-tests/camel-jms/with-external)?

**Comment by zhfeng** [comment]
I think I never test it with `@Identifier("externally-defined")`. So it could be good to add a test with [CustomConnectionFactory](https://github.com/apache/camel-quarkus/blob/main/integration-tests/jms-artemis-client/src/main/java/org/apache/camel/quarkus/component/jms/artemis/it/CustomConnectionFactory.java) by using `@Identifier`. Would you like to work on it?

**Comment by turing85** [comment]
I do no t fully get what the actual gap in coverage is. Do you want to discuss this in the [quarkus zulip chat](https://quarkusio.zulipchat.com/)?

**Comment by turing85** [comment]
On a side note: in the `with-external`, the `@Identifier("externally-defined")` is to activate the health endpoint, nothing more. As you can see in the application code itself, the identifier is not really used anywhere.

**Comment by zhfeng** [comment]
Ok, now I understand that `@Identifier` is only used for health check. If this is the case, I think the current tests in `camel-quarkus` are fine. So this PR could be closed or is there any other thing you need to be concerned with Camel?

**Comment by turing85** [comment]
@zhfeng from our side (`quarkus-artemis`), I think we are fine. It would be nice to have some additional tests on the `camel-quarkus`-side. I'll provide some tests to `camel-quarkus` as soon as I have some spare time. I removed the `@Identifier`-annotation from the `with-external` camel test to make it clear that this annotation is not needed in this case for the feature to work.

**Comment by turing85** [comment]
I was able to validate the current implementation against the relevant camel integration tests (the client-, `sjms`-, and `sjms2`-tests)

**Comment by turing85** [comment]
I took the liberty and re-organized the tests, introducing `common`-modules to eliminate code duplication. The PR "looks" big, but most of it is integration test implementation and refactoring.

File: jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import io.quarkus.artemis.jms.runtime.camel.CamelContextEnhancer;
+import io.quarkus.deployment.Capabilities;
+import io.quarkus.deployment.annotations.BuildStep;
+
+public class CamelJmsRegistryEnhancerProcessor {
+    @SuppressWarnings("unused")
+    @BuildStep
+    AdditionalBeanBuildItem addCamelContextEnhancer(Capabilities capabilities) {
+        if (capabilities.isPresent(CamelCapabilities.CORE)) {
+            return AdditionalBeanBuildItem.unremovableOf(CamelContextEnhancer.class);
```
Comment (zhfeng) [inline jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: I think `CamelContextEnhancer` is not necessary here and `camel-quarkus` should handle it already. Is it possible to remove this class and run the `camel-jms` tests to see if they are still working?

File: jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import io.quarkus.artemis.core.runtime.ArtemisUtil;
+
+@ApplicationScoped
+public class CamelContextEnhancer {
+    void startUp(@Observes CamelContextInitializingEvent event) {
+        Map<String, ConnectionFactory> connectionFactoryNamesFromArc = ArtemisUtil
+                .extractIdentifiers(ConnectionFactory.class, Collections.emptySet());
+        for (var entry : connectionFactoryNamesFromArc.entrySet()) {
+            event.getContext().getRegistry().bind(entry.getKey(), entry.getValue());
```
Comment (zhfeng) [inline jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java:20]: `camel-quarkus` should register these class automatically.

Comment (turing85) [reply jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java:20]: You mean the beans? They are generated with `@Identifier(...)`, thus they are not registered in the context. I tried adding the `@Named(...)`-annotation, but this did not help either.

File: jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java (hunk 1 above)
Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: Do you men removing the whole build step (i.e. the package `io.quarkus.artemis.jms.[deployment|runtime].camel`?

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: Removing class `CamelJmsRegistryEnhancerProducer` results in failing integration tests for module `quarkus-integration-test-artemis-camel-jms-with-default-and-named`:

```
...
Caused by: org.apache.camel.NoSuchBeanException: No bean could be found in the registry for: <default> of type: javax.jms.ConnectionFactory
...
```

Adding `@Named(...)`-annotations to the synthetic beans in addition to removing class `CamelJmsRegistryEnhancerProducer` also results in failing integration tests for module `quarkus-integration-test-artemis-camel-jms-with-default-and-named`

```
...
Caused by: org.apache.camel.NoSuchBeanException: No bean could be found in the registry for: <default> of type: javax.jms.ConnectionFactory
...
```

I have attached [patches (`.zip`)]((https://github.com/quarkiverse/quarkus-artemis/files/10208625/patches.zip)) for both versions if you want to take a look yourself. So as far as I can see, there is no automatism on camel's end, hence we have to add the beans to the camel context manually.

File: jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java (hunk 2 above)
Comment (turing85) [reply jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java:20]: @zhfeng Please see [this comment of mine](https://github.com/quarkiverse/quarkus-artemis/pull/122#discussion_r1045858434).

File: jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java (hunk 1 above)
Comment (zhfeng) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: `<default>` should be the default bean of `ConnectionFactory` ? I'm not sure why you need `from(jms("queue:in").connectionFactory("<default>"))`, just to try `from(jms("queue:in"))` to see if it is working?

Comment (zhfeng) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: As far I as know, it's no need to use `connectionFactory("<default>"))` if there is a defautl ben of `ConnectionFactory`

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: Will try, but I think that the behaviour with not providing a connection factory only works if exactly one connection factory is found.

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: @zhfeng removing `.connectionFactory("<default>")` ([Patch (`.zip`)](https://github.com/quarkiverse/quarkus-artemis/files/10209889/removeConnectionFactory.zip)) leads to `quarkus-integration-test-artemis-camel-jms-with-default-and-named` failing:

```
...
Caused by: java.lang.IllegalArgumentException: connectionFactory must be specified
...

```

File: jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java (hunk 2 above)
Comment (turing85) [reply jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java:20]: @zhfeng can this discussion be resolved?

File: jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java (hunk 1 above)
Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: @zhfeng can this discussion be resolved?

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: @zhfeng can this discussion be resolved?

File: jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java (hunk 2 above)
Comment (turing85) [reply jms/runtime/src/main/java/io/quarkus/artemis/jms/runtime/camel/CamelContextEnhancer.java:20]: I tried some additional things:
- adding `unremoveable()` to the synthetic `ConnectionFactory`-beans has no effect on the current behaviour
- adding dependency `camel-quarkus-bean` has no effect on the current behaviour

**Review by middagj — COMMENTED** [review]
Great work, really becoming a nice extension now that behaves as you would expect

File: docs/modules/ROOT/pages/index.adoc [hunk 3]
```diff
@@ -228,7 +272,7 @@ XA Transactions  must be activated on a per-configuration basis, they are not en
...
 
 == URL as optional configuration
-The `url` configuration property is optional. But in general, without a `url` property, we cannot create a connection. In case of tests with embedded resources or devservices, the corresponding annotation/service injects a url. But if  the application starts up and no `url` property is found, the creation of the bean will throw a `RuntimeException` with a corresponding error message. Notice that this will happen when the bean is created. If the bean is never used (and thus never created), the exception will not occur.
+The `url` configuration property is optional. But in general, without a `url` property, we cannot create a connection. In case of tests with embedded resources or devservices, the corresponding annotation/service injects an url. But if  the application starts up and no `url` property is found, the creation of the bean will throw a `RuntimeException` with a corresponding error message. Notice that this will happen when the bean is created. If the bean is never used (and thus never created), the exception will not occur.
```
[the remark this answers is not in the document]
Comment (turing85) [reply docs/modules/ROOT/pages/index.adoc]: fixed.

File: jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java (hunk 1 above)
Comment (zhfeng) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: @turing85 It's weird and I will try to figure it out. But I'm in holiday these weeks, can you hold this PR and wait for my checking?

Thanks!

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:15]: @zhfeng of course, no problem. Will hold the PR.

Comment (turing85) [reply jms/deployment/src/main/java/io/quarkus/artemis/jms/deployment/camel/CamelJmsRegistryEnhancerProcessor.java:18]: @zhfeng [as discussed](https://github.com/quarkiverse/quarkus-artemis/pull/125#discussion_r1060581213), I added the `name(...)` s and the toggle for the `CamelContextEnhancer`. Do you thin this PR is ready to be merged?

**Review by zhfeng — APPROVED** [review]
Thanks @turing85 and it is a great work!

**Pull request merged**